app/jobs/timeline_collection.py

```diff
@@ -0,0 +1,171 @@
+
+
+import os
+from time import sleep
+from functools import lru_cache
+from dotenv import load_dotenv
+#from tqdm import tqdm as progress_bar
+
+from app import seek_confirmation
+from app.bq_service import BigQueryService, generate_timestamp
+from app.twitter_service import TwitterService
+from app.tweet_parser import parse_timeline_status
+
+load_dotenv()
+
+USER_LIMIT = os.getenv("USER_LIMIT", default="250")
+STATUS_LIMIT = os.getenv("STATUS_LIMIT", default="10_000")
+
+
+class TimelineCollectionJob():
+    def __init__(self, bq_service=None, twitter_service=None, user_limit=USER_LIMIT, status_limit=STATUS_LIMIT):
+        self.bq_service = bq_service or BigQueryService()
+        self.twitter_service = twitter_service or TwitterService()
+
+        self.dataset_address = self.bq_service.dataset_address
+        self.user_limit = int(user_limit)
+        self.status_limit = int(status_limit)
+
+        self.parse_status = parse_timeline_status
+
+        print("---------------------------")
+        print("JOB: TIMELINE LOOKUPS")
+        print("DATASET:", self.dataset_address.upper())
+        print("USER LIMIT:", self.user_limit)
+        print("STATUS LIMIT:", self.status_limit)
+        print("---------------------------")
+
+    def fetch_users(self):
+        sql = f"""
+            SELECT
+                ul.user_id
+                ,prev.latest_status_id -- can be null
+                ,prev.latest_lookup_date -- can be null
+                --,coalesce(prev.latest_lookup_date, '1970-01-01')
+            FROM (
+                SELECT DISTINCT user_id, error_code, follower_count, friend_count, listed_count, status_count, latest_status_id
+                FROM `{self.dataset_address}.user_lookups`
+                WHERE error_code IS NULL and status_count > 0
+                -- LIMIT 10
+            ) ul
+            JOIN (
+                SELECT DISTINCT user_id --, error_type, error_message
+                FROM `{self.dataset_address}.timeline_lookups`
+                WHERE error_type IS NULL
+                -- LIMIT 10
+            ) tl ON ul.user_id = tl.user_id
```

Comment:
this subquery needs to be changed to look at the latest lookup for each user.

Reply:
```sql

SELECT  
  error_type
  ,error_message
  ,count(distinct user_id) as user_count
  ,count(distinct concat(user_id, "-", string(start_at))) as lookup_count
FROM `tweet-collector-py.disinfo_2021_production.timeline_lookups`
GROUP BY 1,2

```


error_type | error_message | user_count | lookup_count
-- | -- | -- | --
  |   | 24904 | 24904
TweepError | Twitter error response: status code = 401 | 1746 | 1746
TweepError | Twitter error response: status code = 503 | 7 | 7
TweepError | Failed to parse JSON payload: Unterminated string starting at: line 1 column 610997 (char 610996) | 1 | 1
TweepError | Failed to send request: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer')) | 13 | 13
TweepError | Failed to send request: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer')) | 7 | 7